relay-server/src/actors/project_cache.rs

```diff
@@ -468,7 +533,60 @@ impl ProjectCacheBroker {
         } = message;
 
         self.get_or_create_project(project_key)
-            .update_state(state, no_cache)
+            .update_state(state.clone(), no_cache);
+
+        // Envelopes need to remain in the queue while Relay receives invalid states from upstream.
+        if state.invalid() {
+            return;
+        }
+
+        let envelopes = self.pending_envelopes.dequeue(&project_key, |queue_key| {
+            let partial_key = if queue_key.own_key == project_key {
+                queue_key.sampling_key
+            } else {
+                queue_key.own_key
+            };
+
+            // We return false if project is not cached or its state is invalid, true otherwise.
+            // We only have to check `partial_key`, because we already know that the `project_key`s `state`
+            // is valid and loaded.
+            self.projects
+                .get(&partial_key)
+                // Make sure we have only cached and valid state.
+                .and_then(|p| p.valid_state())
+                .map_or(false, |s| !s.invalid())
+        });
+
+        // Flush envelopes where both states have resolved.
+        for (envelope, envelope_context) in envelopes {
+            // If the incoming state is the update for the sampling state, we have to get first own
+            // project state and use it for processing the envelope.
+            let own_key = envelope.meta().public_key();
+            let (own_state, sampling_state) = if project_key == own_key {
+                (
+                    Some(state.clone()),
+                    utils::get_sampling_key(&envelope)
+                        .and_then(|key| self.projects.get(&key))
+                        .and_then(|p| p.valid_state()),
+                )
+            } else {
+                (
+                    self.projects.get(&own_key).and_then(|p| p.valid_state()),
+                    Some(state.clone()),
+                )
+            };
```

Comment:
A new addition is here, made in [105b54c](https://github.com/getsentry/relay/pull/1907/commits/105b54cd46dfb225619f506a6ab2a3579a349d8d) - basically if my thinking is correct, we always have to use the own (root) project state for processing

This was not the case, and if the sampling project state update comes in first, after getting all the envelopes for it, we would use this dynamic project state in 2 places.

Reply:
\cc @jjbayer if you could, please, verify my thinking here.

Reply:
Yes, good catch! I remember we fixed this for the project key but apparently we had a similar bug for the project state. We need more tests with `sampling_key != own_key`.

I think what would simplify this code and make it safer against confusion would be to change the signature of `handle_processing` to `fn handle_processing(envelope, envelope_context)`:

* `handle_processing` already does `self.projects.get_mut(project_key)` internally, So we don't gain much from passing `own_state` into it.
* and then we move `utils::get_sampling_key(&envelope).and_then(|key| self.projects.get(&key))` into `handle_processing` as well.

Reply:
Great suggestions!
Addressed this in [0074eb3](https://github.com/getsentry/relay/pull/1907/commits/0074eb372507a488b03907756eb48e003f1c47d4)